### PR TITLE
`FileCache`: add generic support for all datasets

### LINKED ISSUE #1544: Ideas for generic `CachedFile` support across all datasets
WDYT about adding a global `FileCache` to RETURNN and passing every dataset kwargs through `FileCache.handle_cached_files_in_config` in `init_dataset`?

The support for this would also have to ensure we de-register the relevant files when a dataset is de-initialized. I'm not yet sure how to best do this generically. Maybe we can remember the cached files we have gotten from a dataset and then store them in the dataset somewhere for deregistration in `__del__`?

cc @albertz @michelwi @Icemole

### LINKED ISSUE #1549: `DistributeFilesDataset`: copying files blocks `init_seq_order`
When `DistributeFilesDataset.init_seq_order` is called, it will schedule the copying (in the sub proc `_FileCacheProc`), but it will wait for this to complete, thus this will block. But there is no reason for this to block. `init_seq_order` could directly return. Even the `init_seq_order` on the sub dataset can run async.

PR #1546 actually partially fixes this.

### PR DESCRIPTION
**Pull request by NeoLegends**
* Fix #1544.
* Fix #1549.
* Closes #1545

Let me know your thoughts about the implementation here. I'm not a big fan yet about having to register "from the outside" the files to be released later, but I'm also not sure if there's a better way to unregister the files when the dataset is deinitialized.

One advantage realising support this way is that cache operations for the preloaded subepoch in `DistributeFilesDataset` are now actually happening asynchronously, because `init_dataset`, which invokes caching, is called in the subprocess. Previously this caching would happen in the parent process, blocking training.

### CODE REVIEW

File: returnn/datasets/distrib_files.py [hunk 1]
```diff
@@ -324,9 +306,7 @@ def _get_sub_dataset_dict(self, files: List[FileTree]) -> Tuple[Dict[str, Any],
                 f"{self}: sub dataset should have explicit seq_ordering "
                 f"(or seq_order_control_dataset for MetaDataset), got: {dataset_dict}"
             )
-        self._lazy_init_file_cache_proc()
-        dataset_dict, exit_hook = self._file_cache.handle_cached_files_in_config(dataset_dict)
```
Comment (NeoLegends) [inline returnn/datasets/distrib_files.py:328]: This is where the existing code would block the master process.

File: returnn/util/file_cache.py [hunk 2]
```diff
@@ -53,11 +103,11 @@ class FileCache:
...
-        cache_directory: str = "$TMPDIR/$USER/returnn/file_cache",
-        cleanup_files_always_older_than_days: float = 31.0,
-        cleanup_files_wanted_older_than_days: float = 7.0,
-        cleanup_disk_usage_wanted_free_ratio: float = 0.2,  # try to free at least 20% disk space
-        num_tries: int = 3,  # retry twice by default
+        cache_directory: str = DEFAULT_CACHE_DIR,
+        cleanup_files_always_older_than_days: float = DEFAULT_CLEANUP_FILES_ALWAYS_OLDER_THAN_DAYS,
+        cleanup_files_wanted_older_than_days: float = DEFAULT_CLEANUP_FILES_WANTED_OLDER_THAN_DAYS,
+        cleanup_disk_usage_wanted_free_ratio: float = DEFAULT_CLEANUP_DISK_USAGE_WANTED_FREE_RATIO,
+        num_tries: int = DEFAULT_NUM_TRIES,
```
Comment (albertz) [inline returnn/util/file_cache.py:110]: I don't see the benefit in rewriting the code like this. This just makes it unnecessarily more complicated. I would keep the old code.

File: returnn/util/file_cache.py [hunk 3]
```diff
@@ -19,9 +19,59 @@
...
+
+
+DEFAULT_CACHE_DIR = "$TMPDIR/$USER/returnn/file_cache"
+DEFAULT_CLEANUP_FILES_ALWAYS_OLDER_THAN_DAYS = 31.0
+DEFAULT_CLEANUP_FILES_WANTED_OLDER_THAN_DAYS = 7.0
+DEFAULT_CLEANUP_DISK_USAGE_WANTED_FREE_RATIO = 0.2  # try to free at least 20% disk space
+DEFAULT_NUM_TRIES = 3  # retry twice by default
+
+
+cache_instance: Optional["FileCache"] = None
```
Comment (albertz) [inline returnn/util/file_cache.py:35]:
```suggestion
cache_instance: Optional[FileCache] = None
```

File: returnn/util/file_cache.py [hunk 4]
```diff
@@ -19,9 +19,59 @@
...
+
+
+def init_by_config(config: Config):
+    """Initializes the global file cache using values from `config`."""
+    global cache_instance
+    if cache_instance:
+        return
+    # TODO: just pass along all `file_cache_*` config values as kwargs?
```
Comment (albertz) [inline returnn/util/file_cache.py:43]: Yes of course. There should not be single config option for every single arg, but instead just sth like `file_cache_opts`.

File: returnn/datasets/distrib_files.py [hunk 5]
```diff
@@ -647,95 +624,3 @@ def _get(seq_idx: int) -> Optional[DatasetSeq]:
                 raise Exception(f"unknown msg {msg!r}")
     except KeyboardInterrupt:  # when parent dies
         pass
-
-
-class _FileCacheProc:
```
[the remark this answers is not in the document]
Comment (NeoLegends) [reply returnn/datasets/distrib_files.py:652]: See https://github.com/rwth-i6/returnn/issues/1544#issuecomment-2180485412. I'm flexible on this though. Currently (current master) we block the main thread for caching anyways.

File: returnn/util/file_cache.py [hunk 6]
```diff
@@ -331,6 +332,17 @@ def _check_existing_copied_file_maybe_cleanup(src_filename: str, dst_filename: s
...
 
 
+def get_instance(config: Optional[Config] = None) -> FileCache:
+    """
+    Returns a file cache instance potentially initialized by the global config.
+
+    Uses defaults if no global config is set.
+    """
+    config = config or get_global_config(raise_exception=False)
+    kwargs = config.value("file_cache_opts", {}) if config is not None else {}
```
Comment (albertz) [inline returnn/util/file_cache.py:342]:
```suggestion
    config = config or get_global_config(return_empty_if_none=True)
    kwargs = config.typed_value("file_cache_opts") or {}
```

Comment (NeoLegends) [reply returnn/util/file_cache.py:342]: Ah, nice!

File: returnn/datasets/distrib_files.py (hunk 1 above)
Comment (albertz) [reply returnn/datasets/distrib_files.py:328]: For reference, I made #1549 to keep track of this issue.

But I think we can resolve this thread here.

**Comment by albertz** [comment]
Also, this here is missing some logic to not do the full cleanup scan too frequently, e.g. via touching a `.recent_full_cleanup` file as mentioned [here](https://github.com/rwth-i6/returnn/issues/1544#issuecomment-2180755966).

File: returnn/util/file_cache.py [hunk 7]
```diff
@@ -73,7 +74,6 @@ def __init__(
         self._cleanup_files_wanted_older_than_days = cleanup_files_wanted_older_than_days
         self._cleanup_disk_usage_wanted_free_ratio = cleanup_disk_usage_wanted_free_ratio
         self._touch_files_thread = _TouchFilesThread(cache_base_dir=self.cache_directory)
-        self._touch_files_thread.start()
```
Comment (albertz) [inline returnn/util/file_cache.py:76]: I would simply set `self._touch_files_thread: Optional[_TouchFilesThread] = None`, i.e. not create the thread object at all.

Comment (albertz) [reply returnn/util/file_cache.py:76]: Well ok, maybe not so relevant, maybe your current code is also fine.

Comment (NeoLegends) [reply returnn/util/file_cache.py:76]: If we leave this none we need a check+init everywhere. I think the implementation as it is now is simpler in that regard. Also, shouldn't the thread object itself be cheap, and only allocate memory for the OS thread once started?

Comment (albertz) [reply returnn/util/file_cache.py:76]: > only allocate memory for the OS thread once started?

That was exactly the part I was not sure about. I guess it's also implementation specific to CPython. But probably it should be fine.

File: returnn/util/file_cache.py [hunk 8]
```diff
@@ -381,6 +392,7 @@ def __init__(self, *, interval: float = 1.0, cache_base_dir: str):
         self.files = defaultdict(int)  # usage counter
         self.interval = interval
         self.cache_base_dir = cache_base_dir
+        self._started = False
```
Comment (albertz) [inline returnn/util/file_cache.py:395]: This is an attrib you are overwriting from the underlying `Thread` object. See failing test. I think the underlying `Thread` actually already does exactly this logic?

Comment (NeoLegends) [reply returnn/util/file_cache.py:395]: ohhh damn! the docs say there will be a runtimeerror when you invoke start multiple times

Comment (NeoLegends) [reply returnn/util/file_cache.py:395]:
```
if self._started.is_set():
    raise RuntimeError("threads can only be started once")
```

**Comment by albertz** [comment]
Test case needs to be fixed:
```
  File "/home/runner/work/returnn/returnn/tests/test_Util.py", line 873, in test_file_cache
    line: print("Recent cleanup is ago:", time.monotonic() - cache._recent_full_cleanup_time, "secs")
    locals:
      print = <builtin> <built-in function print>
      time = <global> <module 'time' (built-in)>
      time.monotonic = <global> <built-in function monotonic>
      cache = <local> <returnn.util.file_cache.FileCache object at 0x7f9b7acbe650>
      cache._recent_full_cleanup_time = <local> !AttributeError: 'FileCache' object has no attribute '_recent_full_cleanup_time'
AttributeError: 'FileCache' object has no attribute '_recent_full_cleanup_time'
```

**Review by albertz — APPROVED** [review]

**Comment by albertz** [comment]
I just noticed that now some i6_core tests fail, because this PR here now makes `tree` (`pip install dm-tree`) a mandatory dependency for RETURNN. Before it was optional (only needed when the user used some code where it was used).

So is this ok?

cc @curufinwe @michelwi @NeoLegends @JackTemaki @Atticus1806 @christophmluscher (not sure who else)

**Comment by michelwi** [comment]
The dm-tree package is installed in all of our singularity images since it became an (optional) requirement back then. Also in all cases where returnn is used, a fixed commit is attached to the setup. Therefore non of our old setups should break.

> So is this ok?

yes from my side.

**Pull request merged**